src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,294 @@
+        String enumValue = (String) value;
+        if (!Objects.requireNonNull(nonNullableSchema.getEnumValues()).contains(enumValue)) {
+          throw new UnexpectedFormatException(String.format("Value '%s' is not enum value. Enum values are: '%s'",
+                                                            enumValue, nonNullableSchema.getEnumValues()));
+        }
+
+        return enumValue;
+      default:
+        throw new UnexpectedFormatException("Field type '" + fieldType + "' is not supported.");
+    }
+  }
+
+  private DBObject extractMap(Object value, Schema schema) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    Map.Entry<Schema, Schema> mapSchema = Objects.requireNonNull(schema.getMapSchema());
+    if (mapSchema.getKey().getType() != Schema.Type.STRING) {
+      throw new UnexpectedFormatException("Key of the 'map' schema must be of the 'string' type.");
+    }
+    if (!(value instanceof Map)) {
+      throw new UnexpectedFormatException(String.format("Unexpected value of 'map' type: '%s'", value));
+    }
+
+    Map<String, ?> map = (Map<String, ?>) value;
+    BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
+    map.forEach((k, v) -> bsonBuilder.add(k, extractValue(v, mapSchema.getValue())));
+
+    return bsonBuilder.get();
+  }
+
+  private List extractArray(Object value, Schema schema) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    if (!(value instanceof List)) {
+      throw new UnexpectedFormatException(String.format("Unexpected value of 'array' type '%s'", value));
+    }
+    List values = (List) value;
+    List<Object> extracted = new ArrayList<>();
+    Schema componentSchema = schema.getComponentSchema();
+    for (Object obj : values) {
+      extracted.add(extractValue(obj, componentSchema));
+    }
+    return extracted;
+  }
+
+  private Object extractUnion(Object value, Schema schema) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+
+    List<Schema.Type> unionTypes = schema.getUnionSchemas().stream()
+      .map(s -> s.isNullable() ? s.getNonNullable() : s)
+      .map(Schema::getType)
+      .collect(Collectors.toList());
+    // Extract values of types that can not be written directly or may contain nested values of such types
+    if (unionTypes.indexOf(Schema.Type.BYTES) != -1 && value instanceof ByteBuffer) {
+      // ByteBuffer -> byte[] conversion required
+      int schemaIndex = unionTypes.indexOf(Schema.Type.BYTES);
+      return extractValue(value, schema.getUnionSchema(schemaIndex));
+    } else if (unionTypes.indexOf(Schema.Type.ENUM) != -1 && value instanceof String) {
+      // Enum validation required
+      int schemaIndex = unionTypes.indexOf(Schema.Type.ENUM);
+      return extractValue(value, schema.getUnionSchema(schemaIndex));
+    } else if (unionTypes.indexOf(Schema.Type.ARRAY) != -1 && value instanceof List) {
```

Comment:
this is incorrect, an 'array' field can be a Collection or an array.

Reply:
Removed. Implementation changed to:
```
    for (Schema s : schema.getUnionSchemas()) {
      try {
        return extractValue(fieldName, value, s);
      } catch (Exception e) {
        // expected if this schema is not the correct one for the value
      }
```